plugins/woocommerce/client/blocks/assets/js/blocks/add-to-cart-with-options/variation-selector/frontend.ts

```diff
@@ -285,6 +267,46 @@ const { actions, state } = store< VariableProductAddToCartWithOptionsStore >(
 				const matchedVariationId = matchedVariation?.variation_id;
 				productDataActions.setVariationId( matchedVariationId ?? null );
 			},
+			validateVariation() {
+
+				const { state: testState, actions: addToCartWithOptionsActions } = store<AddToCartWithOptionsStore>(
+						'woocommerce/add-to-cart-with-options',
+						{},
+						{ lock: universalLock }
+					);
+
+				addToCartWithOptionsActions.clearErrors( 'variable-product' );
+
+				const { availableVariations, selectedAttributes } = getContext<Context>();
+				const matchedVariation = getMatchedVariation(
+					availableVariations,
+					selectedAttributes
+				);
+
+				const variationId = matchedVariation?.variation_id;
+				const inStock = matchedVariation?.is_in_stock;
+
+				if (!variationId || !Boolean(inStock)) {
+					const { errorMessages } = getConfig();
+					
+					const error = {
+						code: 'variableProductMissingAttributes',
+						message: errorMessages?.variableProductMissingAttributes || '',
+						group: 'variable-product',
+					};
+
+					if ( variationId && !Boolean(inStock) ) {
+						error.code = 'variableProductOutOfStock';
+						error.message = errorMessages?.variableProductOutOfStock || '';
+					}
+
+					if ( error?.message ) {
+						addToCartWithOptionsActions.addError( error );
+					}
+				}
+
```

Comment:
Nit, I would handle this with early returns:

```suggestion
				if ( ! matchedVariation?.variation_id ) {
					actions.addError( {
						code: 'variableProductMissingAttributes',
						message:
							errorMessages?.variableProductMissingAttributes ||
							'',
						group: 'variable-product',
					} );
					return;
				}

				if ( ! matchedVariation?.is_in_stock ) {
					actions.addError( {
						code: 'variableProductOutOfStock',
						message: errorMessages?.variableProductOutOfStock || '',
						group: 'variable-product',
					} );
					return;
				}
```